### ISSUE TITLE
Gather all of the outputs from different nodes and then compute contrastive loss

### ISSUE THREAD

**Issue by BaohaoLiao** [PANXiao1994/mRASP2#11]
Hi,

this is a really great paper. I have a question about the calculation of the contrastive loss.

In the paper, you "use 8 × 4 NVIDIA V100 with update frequency 50 to train the models and each batch contains about 3 million tokens". Do you compute the contrastive loss within a mini-batch on a GPU or on all GPUs? 

For the conventional contrastive loss computing, we need to gather all outputs from the models on different GPUs, and then compute it. But in your code, I can not find the code for that. So do you only compute it with the outputs only on a GPU? 

If I misunderstand your code, could you point out where you gather the outputs? Normally we use something like "torch.gather_all()".

**Comment by PANXiao1994** [comment]
We compute the contrastive loss within a mini-batch on a GPU. Then we average the contrastive loss across batches (along with cross entropy loss). This is an approximation and saves the memory and network cost.
